docs/design/2020-09-28-cluster-read-only-mode.md

```diff
@@ -0,0 +1,243 @@
+# Proposal: Support Global Read Only Mode
+
+- Author(s):    [ichn-hu](https://github.com/ichn-hu)
+- Last updated:  Oct. 09, 2020
+- Discussion at:  
+
+## Abstract
+
+This document describes the design of the `read_only` global variable that once turned on blocks non-super users from updating the TiDB cluster.
+
+
+## Motivation or Background
+
+As requested and described in [this issue](https://github.com/pingcap/tidb/issues/18426), such a variable can be beneficial to maintenance, migration and replication, and according to a [blog post](https://pingcap.com/blog/heterogeneous-database-replication-to-tidb#creating-a-read-only-or-archive-database), TiDB can be used as a read-only or archive database, in which case we would like to make the cluster running in a read only mode, while only allowing CDC tools to update the data.
+
+
+## Detailed Design
+
+Based on the issue description and [MySQL’s description](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_read_only), the `read_only` option of TiDB will act similarly.
+
+
+### Behavior Description
+
+
+*   `read_only` is disabled by default.
+*   The only way to enable `read_only` mode is to execute `set global read_only = on` with a SuperPriv user.
+    *   The enabling attempt will block until the whole cluster enters the `read_only` mode.
+    *   And only `set global read_only = off` to turn off the mode.
+    *   The `read_only` global variable will have 3 values, the reason for such design is to make the status more intuitive, see section technical issues 4 for a detailed explanation.
+        *   off, when the cluster acts in a normal mode
+        *   enabling, when the turning on process is going on
+        *   on, when the whole cluster is in read only mode
+    *   `set global read_only = on` upon return
+        *   if there are warnings, meaning there are still some TiDB servers in the cluster that are possibly not read only, the user should check if these TiDB servers are down or have unreleased write lock or have large transactions committing.
+            *   The warnings contain all known TiDB servers that have not acknowledged the read only mode.
+            *   If the user examines the `read_only` global variable, it will be `enabling`, not `on`.
+            *   She could rerun `set global read_only = on` to wait for another period of time (default to 10 seconds, could have exponential backoff behavior).
+            *   If some of the TiDB servers reported in the warnings are indeed down, the user should remove them and rerun the enabling command, and expect an immediate return of success.
+        *   if there is another session running `set global read_only = off` while waiting for enabling, an error will be returned, marking the fail of the attempt. \
```

Comment:
> 1. IMO, this attempt can be idempotent. Is that correct?

Yes.

> 2. What if another session is in another tidb-server instance?

That doesn't matter. The polling strategy will also examine the global variable, and if it is being set to off, the polling process will be aborted.

Reply:
> The alternative behavior is for the subsequent sessions to block and wait for the state to change to `!= 'enabling` before they are permitted to do anything.
> 
> I would hope that there are not scripts that are concurrently trying to set the server into `read_only`, but incorporating internal retry would be more compatible.

I think allowing setting off is valuable if the user want to cancel the enabling of read only.

Since enabling attempt is idemopotent, it dosen't matter if there are multiple sessions want to enable read only -- they just turn the read only variable from on to enabling and poll until either all servers reported or the variable changed.